### ISSUE TITLE
Forcing a specific Gateway inside a Firewall rule stopped working properly in 17.1 if the Gateway is associated with an OpenVPN Interface

### ISSUE THREAD

**Issue by mmssantos** [opnsense/core#1378]
Upgraded to 17.1 this morning and my configured scenario stops working properly.

I have several OpenVPN tunnels in client mode that provide additional Internet access paths to OPNsense.

Inside the firewall rules, I have several rules forcing several specific traffic flows through the Gateway associated with those OpenVPN tunnels and the setup was working properly up to 16.7.14.

With todays upgrade, Traffic is now not forced through the OpenVPN tunnel, instead going out the Default Gateway.

Even more interesting is the fact that on the Firewall Logs, the triggered rule is the correct rule:

  @118 pass in log quick on igb1 route-to (ovpnc1 xxx.xxx.xxx.xxx) inet from <alias:0> to any flags S/SA keep state label "USER_RULE"

I will try to rebuild the complete setup at the end of the try to see if it can be something related to the migration from 16.7.14 to 17.1 and will provide additional information when finished.

If you guys need any additional debug logs or anything, be sure to let me know.

**Comment by fichtner** [comment]
hi @mmssantos,

please try this kernel:

    # opnsense-update-kr  17.1-route
    # /usr/local/etc/rc.reboot

If that doesn't help see if setting this does the trick:

    # sysctl net.pf.share_forward=0


Thanks,
Franco

**Comment by mmssantos** [comment]
Thank you for the quick reply Franco,

Updating to kernel 17.1-route seems to have done the trick, will do some additional testing and send you feedback as soon as possible.

**Comment by fichtner** [comment]
perfect, thank you so far!

**Comment by mmssantos** [comment]
Morning Franco,

Everything looks to be working like on the previous versions apart from an issue that I am not sure was present on previous versions or not.

I have a Firewall Rule forcing all Traffic from a specific source IP Address through an OpenVPN tunnel and that is working great now with the updated kernel.

Under that Firewall Rule, I have another Rule blocking all Traffic from that IP Address to everywhere, so if the OpenVPN tunnel should go down, all Outbound Traffic from that Source IP Address should get blocked.

However, I tested the scenario yesterday and when the OpenVPN tunnel was down, Outbound Traffic from that Source IP Address was being pushed through the Default Gateway and reviewing the Firewall Logs, the Rule that was allowing the Traffic was the Rule that forced Traffic through the OpenVPN tunnel.

Not sure on the behaviour on the previous versions since I didn't notice this behaviour.
Is this the expected behaviour in general or might it be a new issue with version 17.1?

If you need any additional debug information please be sure to let me know.

**Comment by fichtner** [comment]
Hi Márcio,

It could be new. We tried to be cautious in our approach to the new policy routing, but that may have been too much. If you could confirm this wasn't the case on 16.7, I can change it and provide another test kernel.


Thanks,
Franco

**Comment by mmssantos** [comment]
I can setup a vm with 16.7.14 and validate if that occurs, but it will take me a few days.

Will get back to you as soon as possible.

Thanks for all the help,
Márcio

**Comment by fichtner** [comment]
Ok, let's cut this short... please see if this kernel does the trick :)

# opnsense-update -kr 17.1-route-drop


Cheers,
Franco

**Comment by fichtner** [comment]
Ok here we go :)

    # opnsense-update -kr 17.1-routedrop
    # /usr/local/etc/rc.reboot


Cheers,
Franco

**Comment by samurailink3** [comment]
Thanks @mmssantos  for reporting this and thanks @fichtner for the fix! I also ran into this problem. The 17.1-routedrop kernel did the trick!

**Comment by mmssantos** [comment]
Morning all.

Just upgraded to 17.1.1 and some stuff is broken again.

Haven't changed anything from the baseline packages installed by the upgrade and the symptoms are the following:

  1. Issue with forced traffic via an OpenVPN Tunnel Gateway being allowed through the Default Gateway if the OpenVPN Tunnel is down when there is a rule explicitly blocking it from entering the Default Gateway is back.

  2. Incoming traffic from the OpenVPN Tunnel that triggers a NAT Port Forward rule gets the forward to the internal IP/Port done correctly, however, when the internal IP/Port returns Traffic, it is placed inside the OpenVPN Tunnel without being masqueraded, effectively killing the communication.

Applying the indicated sysctl net.pf.share_forward=0 command:

  1. Issue still remains. OpenVPN Tunnel is down and traffic is being forwarded via the Default Gateway even though that particular traffic stream is blocked with an explicit firewall rule. Analysing the firewall logs, the rule that is being triggered for the traffic flow is the rule that forces traffic through the OpenVPN Tunnel. It appears as though when the Tunnel is down, OPNsense is silently ignoring the forced gateway and on the same rule just forwards traffic through the Default Gateway instead.

  2. Behaviour returns to normal, return traffic is injected inside the tunnel properly masqueraded.

Please let me know if you need any additional logs or testing done.

Thanks,
Márcio

**Comment by fichtner** [comment]
Hi Márcio,

We're fixing this step by step, 17.1.1 did not have all the fixes. Here is a newer kernel based on the droproute changes:

    # opnsense-update -kr 17.1.1-route
    # /usr/local/etc/rc.reboot

If net.pf.share_forward=0 is the same even after a fresh reboot (try setting it through System: Settings: Tunables for persistency after reboot).

We're thankful for all observed behaviour with and without the sysctl setting. :)


Cheers,
Franco

**Comment by mmssantos** [comment]
Thank you for all the support Franco.

With the new Kernel and net.pf.share_forward=0 after a clean reboot:

  Traffic is no longer forced into the OpenVPN Tunnel, instead flowing through the Default Gateway;

  With the OpenVPN Tunnel down and the Firewall Rule blocking all traffic from a specific IP Address through the Default Gateway, traffic flows through the Default Gateway and according to Firewall Logs, the Rule that is getting triggered is the OpenVPN Tunnel forced Gateway.

With the new Kernel and net.pf.share_forward=1 after a clean reboot:

  Traffic is no longer forced into the OpenVPN Tunnel, instead flowing through the Default Gateway;

  With the OpenVPN Tunnel down and the Firewall Rule blocking all traffic from a specific IP Address through the Default Gateway, traffic flows through the Default Gateway and according to Firewall Logs, the Rule that is getting triggered is the OpenVPN Tunnel forced Gateway.

So it looks like this new kernel put us in an even worse state then before :).

Going back to the Default 17.1.1 Kernel with net.pf.share_forward=0 for the time being.

Available for any additional information collection or kernel testing.

Thanks,
Márcio

**Comment by fichtner** [comment]
Hi Márcio,

This is very odd, because net.pf.share_forward=0 is the old default, the one that FreeBSD 11.0 has and hopefully the behaviour that FreeBSD had on 10.3 (16.7).

At this point, the gateway down issue you reported seems to lie beyond our control and we need confirmation how this was on 16.7.

We're considering switching back to net.pf.share_forward=0 with a GUI button to enable the behaviour for people using Multi-WAN and Captive Portal / Traffic Shaper for 17.1.2. That's the best we can do at this moment, before we start chasing down issues within FreeBSD itself, lastly circling back to making net.pf.share_forward=1 fully compatible with FreeBSD.


Cheers,
Franco

**Comment by fichtner** [comment]
Hi Márcio,

We should be back into a working order on 17.1.2. Can you confirm or deny? If something is not working we need to treat it as a separate bug and look for clues in the FreeBSD code, not our patches.


Cheers,
Franco

**Comment by QuentinC** [comment]
Hello Franco,

I just wanted to comment on this because I have the same behavior.

I have one /24 network that is expected to be forced to get out from some VPN. For this, I'm using out NAT, and I am forcing the gateway to the VPN one for this network.

This is working as expected when the VPN is running, but traffic goes to the default gateway if the VPN is down. 

I'm running on 17.1.2.
Right now, I am not able to block traffic from my VPN network to the default gateway..

Thanks,

Quentin

**Comment by fichtner** [comment]
Hi Quentin et al,

So here's what I gather from this: 17.1.2 behaviour is unchanged, gateway does not block if down. As far as I can tell  this is stock FreeBSD 11.0 behaviour at this point, so we have to dive into the commits there to see what changed.

Still, I would like final confirmation that this wasn't problematic on 16.7.x at all?


Thank you,
Franco

**Comment by QuentinC** [comment]
Ok I will try this, but not before the week-end I think.
Anyway, I'm not sure that there isn't a problem with the "down gateway detection". 

If I cut down the VPN link, in my "Gateway status" it is still in green, but without any IP address...

![chrome_2017-03-01_12-01-54](https://cloud.githubusercontent.com/assets/3394508/23457163/0a3fc8e4-fe77-11e6-86fe-4b2c0d71bd1c.png)

Thanks,

Quentin

**Comment by QuentinC** [comment]
Hello Franco,

I've just finished setting up some VM vith 16.7 and configuring te VPN.
I experience exactly the same behavior on it ! (When VPN is down packets are routed thru the default gateway...)

Here is the version used:
```
OPNsense 16.7-amd64
FreeBSD 10.3-RELEASE-p5
OpenSSL 1.0.2h 3 May 2016
```

I will keep this VM if needed fore more testings.

Thanks, 

Quentin

**Comment by fichtner** [comment]
Quentin, thank you very much. I've retagged the scope of this ticket. Would be nice to add so first need to find the code responsible. Now we know where to look: the packet filter itself :)

**Comment by ErAzOr2k** [comment]
I ran into the same issue.
When my VPN goes down, the traffic will be routed through the default gateway. I'm running 17.1.6.
Is there a workaround for this version?

**Comment by fichtner** [comment]
We are back to how 10.3 worked with our 11.0 code, I think there was a misconception cleared up that the previous version didn't do this either.
